internal/planner/telemetry.go

```diff
@@ -0,0 +1,122 @@
+package planner
+
+import (
+	"context"
+	"sync"
+
+	"github.com/canonical/github-runner-operators/internal/telemetry"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+const (
+	flavorPlatform           = "github" // Currently only github is supported
+	flavorPressureMetricName = "mayfly_planner_flavor_pressure"
+)
+
+// must is a helper function that panics if an error is encountered, else returns the object.
+func must[T any](obj T, err error) T {
+	if err != nil {
+		panic(err)
+	}
+	return obj
+}
+
+var (
+	pkg                  = "github.com/canonical/github-runner-operators/internal/planner"
+	meter                = otel.Meter(pkg)
+	logger               = telemetry.NewLogger(pkg)
+	flavorPressureMetric = must(
+		meter.Int64ObservableGauge(
+			flavorPressureMetricName,
+			metric.WithDescription("The current pressure value for each flavor"),
+			metric.WithUnit("{pressure}"),
+		),
+	)
+)
+
+// Metrics encapsulates all OpenTelemetry metrics for the planner service.
+type Metrics struct {
+	mu        sync.RWMutex
+	pressures map[[2]string]int64 // key is [platform, flavor]
+}
+
+// NewMetrics initializes the planner metrics and registers the observable gauge.
+func NewMetrics() *Metrics {
+	m := &Metrics{
+		pressures: make(map[[2]string]int64),
+	}
+
+	must(
+		meter.RegisterCallback(
+			func(_ context.Context, observer metric.Observer) error {
+				m.mu.RLock()
+				defer m.mu.RUnlock()
+
+				for key, value := range m.pressures {
+					platform, flavor := key[0], key[1]
+					observer.ObserveInt64(
+						flavorPressureMetric,
+						value,
+						metric.WithAttributes(
+							attribute.String("platform", platform),
+							attribute.String("flavor", flavor),
+						),
+					)
+				}
+				return nil
+			},
+			flavorPressureMetric,
+		),
+	)
+
+	return m
+}
+
+// ObserveFlavorPressure updates the gauge value for a given platform and flavor.
+func (m *Metrics) ObserveFlavorPressure(platform, flavor string, value int64) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.pressures[[2]string{platform, flavor}] = value
```

Comment:
The problem with caching the pressure value in memory is that there may be multiple instances of the planner server. When a flavor change hits one server, it updates the pressure metric for that server only and does not affect the others.

Instead, I think it would be better to read the pressure value from the database inside the metrics callback, since performance shouldn't be an issue, metrics scraping typically occurs only every 5–60 seconds.